internal/overlord/servstate/manager.go

```diff
@@ -466,6 +475,19 @@ func (m *ServiceManager) Replan() ([]string, []string, error) {
 	return stop, start, nil
 }
 
+// NeedsRestart compares the old and new config structs for a service, to
+// determine if the service needs to be restarted. For example, a service
+// should not be restarted if only the log targets have changed.
+func NeedsRestart(old, new *plan.Service) bool {
+	// TODO: make this more conservative. Changes to metadata, dependencies,
+	// service actions, etc shouldn't require a restart.
+	oldCopy := old.Copy()
+	oldCopy.LogTargets = nil
+	newCopy := new.Copy()
+	newCopy.LogTargets = nil
+	return !reflect.DeepEqual(oldCopy, newCopy)
```

Comment:
The problem we're trying to solve here is that not all changes to `plan.Service` require a service restart. In particular, when we added the `LogTargets` field to `plan.Service`, we noticed in tests that **only** changing the `LogTargets` would trigger a service restart. There's no reason this should happen, since the service **itself** has not changed - only where its logs are being sent, and this is managed independently by the log manager. (Changes to service metadata and some other fields shouldn't require a restart either, but this is not of primary concern right now).

Hence, we created a `NeedsRestart` method that checks if there has been a **fundamental** change to a service which would necessitate a restart. This is more appropriate for the use case (in `Replan`, we we using `service.Equal` to determine whether to restart a service). `NeedsRestart` seems more appropriate inside the `servstate` package, as it's concerned with the semantics of the service config from the POV of the service manager, and the `plan` package should be decoupled from the semantics. The only use of the `service.Equal` method was inside `Replan`, hence we removed the unused method.

If your concerns are more about the implementation of `NeedsRestart`, then fair enough, we can rewrite this so that we are not deep-copying the service structs.

Reply:
We're mostly in agreement, except here:

> NeedsRestart seems more appropriate inside the servstate package, as it's concerned with the semantics of the service config from the POV of the service manager, and the plan package should be decoupled from the semantics.

That doesn't seem to be the case. The analysis being done tells if a service setup has been modified in a way that changes what should be started. This is a static analysis of the plan that can be done completely by itself, with no regards for what the manager or the system state is at all. That was already the case before with `service.Equal`, and it remains the case even if we're tuning its meaning because now there are references to external objects that do not modify how the process is started.

It may look like this:
```
// SameStart returns whether s is started the same way as other.
func (s *Service) SameStart(other *Service) bool { ... }
```